[FSDP2] Required `mesh_dim_names` for HSDP

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137436

Two changes:
1. Require `mesh_dim_names` if using HSDP
2. Pass only the shard mesh to `fsdp_pre_all_gather`

Change 1 is technically BC breaking, but it should not be hard to fix on the user side.

cc @XilunWu @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Commit:
[FSDP2] Required `mesh_dim_names` for HSDP

[ghstack-poisoned]

awgu commented:
should add an HSDP test for all-gather extensions

Commit:
Update on "[FSDP2] Required `mesh_dim_names` for HSDP"




cc XilunWu H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Required `mesh_dim_names` for HSDP"




cc XilunWu H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu commented:
@pytorchbot merge

Skylion007 commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -753,6 +753,16 @@ def _get_grad_inner_tensor(self, grad: torch.Tensor) -> torch.Tensor:
+            return mesh
+        elif mesh.ndim == 2:
+            assert mesh.mesh_dim_names is not None
+            return mesh[mesh.mesh_dim_names[-1]]
```
Doesn't this mean 3D parallelsim won't be supported like it is for FSDP1?

malfet commented:
@pytorchbot revert -m "Looks like it broke distributed testing, see https://github.com/pytorch/pytorch/actions/runs/11239761070/job/31249854217" -c nosignal

awgu commented:
I think I just missed one `init_device_mesh` callsite:

```
AssertionError: Please init the 2D mesh for HSDP with mesh_dim_names specified

To execute this test, run the following from the base repo dir:
    python test/distributed/_composable/fsdp/test_fully_shard_training.py TestFullyShardGradientAccumulation.test_gradient_accumulation
```

let me fix this, run CI, and try to reland tomorrow

Commit:
Update on "[FSDP2] Required `mesh_dim_names` for HSDP"


Two changes:
1. Require `mesh_dim_names` if using HSDP
2. Pass only the shard mesh to `fsdp_pre_all_gather`

Change 1 is technically BC breaking, but it should not be hard to fix on the user side.

cc XilunWu H-Huang kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu commented:
alright let us try this again
ran both periodic and inductor CI, so this should cover all distributed tests

awgu commented:
@pytorchbot merge
